src/Dsn.php

```diff
@@ -0,0 +1,228 @@
+            if (!isset($parsedDsn[$component]) || (isset($parsedDsn[$component]) && empty($parsedDsn[$component]))) {
+                throw new \InvalidArgumentException(sprintf('The "%s" DSN must contain a scheme, a host, a user and a path component.', $value));
+            }
+        }
+
+        if (isset($parsedDsn['pass']) && empty($parsedDsn['pass'])) {
+            throw new \InvalidArgumentException(sprintf('The "%s" DSN must contain a valid secret key.', $value));
+        }
+
+        /** @psalm-suppress PossiblyUndefinedArrayOffset */
+        if (!\in_array($parsedDsn['scheme'], ['http', 'https'], true)) {
+            throw new \InvalidArgumentException(sprintf('The scheme of the "%s" DSN must be either "http" or "https".', $value));
+        }
+
+        /** @psalm-suppress PossiblyUndefinedArrayOffset */
+        $segmentPaths = explode('/', $parsedDsn['path']);
+        $projectId = array_pop($segmentPaths);
+
+        if (!ctype_digit($projectId)) {
+            throw new \InvalidArgumentException('"%s" DSN must contain a valid project ID.');
+        }
+
+        $lastSlashPosition = strrpos($parsedDsn['path'], '/');
+        $path = $parsedDsn['path'];
+
+        if (false !== $lastSlashPosition) {
+            $path = substr($parsedDsn['path'], 0, $lastSlashPosition);
+        }
+
+        /** @psalm-suppress PossiblyUndefinedArrayOffset */
+        return new self(
+            $parsedDsn['scheme'],
+            $parsedDsn['host'],
+            $parsedDsn['port'] ?? ('http' === $parsedDsn['scheme'] ? 80 : 443),
+            (int) $projectId,
+            $path,
+            $parsedDsn['user'],
+            $parsedDsn['pass'] ?? null
+        );
+    }
+
+    /**
+     * Gets the protocol to be used to access the resource.
+     */
+    public function getScheme(): string
+    {
+        return $this->scheme;
+    }
+
+    /**
+     * Gets the host that holds the resource.
+     */
+    public function getHost(): string
+    {
+        return $this->host;
+    }
+
+    /**
+     * Gets the port on which the resource is exposed.
+     */
+    public function getPort(): int
+    {
+        return $this->port;
+    }
+
+    /**
+     * Gets the specific resource that the web client wants to access.
+     */
+    public function getPath(): string
+    {
+        return $this->path;
+    }
+
+    /**
+     * Gets the ID of the resource to access.
+     */
+    public function getProjectId(): int
+    {
+        return $this->projectId;
+    }
+
+    /**
+     * Gets the public key to authenticate the SDK.
+     */
+    public function getPublicKey(): string
+    {
+        return $this->publicKey;
+    }
+
+    /**
+     * Gets the secret key to authenticate the SDK.
+     */
+    public function getSecretKey(): ?string
+    {
+        return $this->secretKey;
+    }
+
+    /**
+     * Gets the URL of the API endpoint to use to post an event to Sentry.
+     */
+    public function getApiEndpointUrl(): string
```

Comment:
Maybe we want to call this something like `getEventApiEndpointUrl` or even `getEventStoreApiEndpointUrl` since this method name seems a bit too generic for the specific endpoint it returns.

You might want to add a `$path` paramater to this method and add the `getEventStoreApiEndpointUrl` method too which call this one with a path (or leave that method out and call this one with a path from the http transport. This so we can use it to "generate" other API endpoints too. 

Little bit of future proofing for upcoming features we need to call API endpoints for.

Reply:
I decided to call the method `getStoreApiEndpointUrl`, based on some similarity with what has been done in the Go SDK. 

>You might want to add a $path paramater to this method

I don't think that it's worth adding it now that we don't even have a case to justify it, plus if we will ever have the need of generating more endpoint URLs then it will probably be better to rely on separate methods.

Reply:
Fair enough 👌 Good change!